### ISSUE TITLE
Download files for offline access

### ISSUE THREAD

**Issue by kojid0** [cryptomator/ios#144]
### Please agree to the following

- [X] I have searched [existing issues](https://github.com/cryptomator/ios/issues?q=) for duplicates
- [X] I agree to follow this project's [Code of Conduct](https://github.com/cryptomator/ios/blob/develop/.github/CODE_OF_CONDUCT.md)

### Summary

Add an option to download files/folders for offline access

### Motivation

In case of slow or no internet, a function to access the files offline is very handy. The files should be encrypted when the vault is locked.

**Comment by idcristi** [comment]
I have the same problem. I downloaded the files to access them offline, but the problem is that the files are loading and not working. It would be great to have such a feature to access the files in offline mode. I checked on the desktop version, and it works without problems.

**Comment by Marcusg562** [comment]
I think this would be the way:

Motivation
Vaults are mostly stored online which is quite nice because you've on all device the same state. But while on traveling (or anywhere else in the world with a bad connection) you can't access them. It would be nice if you cold add an option for a local stored copy of the online vault. Which is updating frequently or on demand (best like an rsync ;) ).

Considered Alternatives
My actual doing. Deleting the local vault, copy the existing vault to local. But since this is a manual doing it always takes time.

**Comment by Marcusg562** [comment]
@tobihagemann how would be without a sync and just a download/copy within the App? So for example you add a button in the app for a manual copy to the phone. You can specify an additional offline storage in the app. Especially on iOS you could choose between your Phone and the iCloud. So each Vault you choose will be located in this folder. In case you wanna update the local fault, you perform a delete of the files and download it new again. Not a real sync but at least a possibility to download selected vaults easily.

**Comment by nferch** [comment]
Curious what sort of support there is under the hood for the cloud providers.

For example, I can choose to "Make available offline" files in Dropbox, but I get a "Content unavailable" error when trying to open files on a Cryptomator vault when I don't have Internet.

iCloud Drive "just works", allowing me to open and unlock a vault when offline and will upload the changes when Internet connectivity is restored.

**Comment by nferch** [comment]
[...]

> Actually if some other cloud provider like Dropbox decides to support folder enumeration with their FileProviderExtension in the futures it would be the same. But to be honest I don't think this will happen since this API was not really designed for cloud providers (and is also completely undocumented!).
> 
> Therefore, I still think we need to handle this some time in the future but as @tobihagemann already said it's quite complex to get this right.

Makes sense, definitely appreciate the complexity in figuring this out, both from the UX and implementation standpoint.

FWIW my use cases are currently very small files, so having the app syncing a tarball of a Vault seems like it would work, as other apps that handle single files with Dropbox seem to work ok.

**Comment by wittymap** [comment]
> iCloud Drive "just works", allowing me to open and unlock a vault when offline and will upload the changes when Internet connectivity is restored.
@nferch - How were you able to get iCloud Drive to "just work", out of curiosity? 

I have a vault in iCloud Drive and downloaded the full vault through the iCloud section of the files app while online. However when later offline and trying to view a file in the Cryptomator section of the files app, an error dialog saying "You're offline" prevents me from viewing any files (even though I downloaded the full vault already through the iCloud section) and can open the encrypted files in the iCloud section of the files app.

**Comment by nferch** [comment]
@wittymap I suspect the difference between our usages is that I created the vault using the iOS app, choosing iCloud drive as the storage. So it appears with the cloud icon in the Cryptomator vault list.

I can access the decrypted files in the Files app under "Locations", which shows the Cryptomator vault, which in turn provides access to the files.

**Comment by soberdor** [comment]
It’s a pity this seems to be such a hard nut to crack. To me it appears to be a critical feature. While read-only would already be a very nice step forward, even a two-way sync would seem worthwile pursuing (I would be willing to pay for such an upgrade). However being off-network (eg on a plane or a ferry) and not having access to any files at all is a total bummer for everyday use.

**Comment by ewerspej** [comment]
This should be a first-class feature in any Cloud-based software. Boxcryptor had this option, this must be implement ASAP in Cryptomator or it's quite useless. Cloud storage and encryption cannot require permanent, fast internet access.

**Comment by mythofechelon** [comment]
I'd like this too.

**Comment by fiore42** [comment]
Happy New Year! (for some)

I just switched to Cryptomator and I'd like to hear if there is any plan to support this.

**Comment by ewerspej** [comment]
@tobihagemann Any update on this? After all, Cryptomator already has a caching mechanism. Can offline access not build on top of that? While offline access is less important on mobile phones, it certainly is important on desktop computers in case of network outages, etc.

**Comment by fiore42** [comment]
@ewerspej funny... I have the opposite opinion. 

You don't need this option for desktop, since google drive has a "make available offline" option. When you select this for a specific folder, google drive will keep a local copy on your disk.
on the other side, a cache is necessary on mobile if you find yourself offline. No data in a foreign country, out of coverage, etc.

**Comment by ewerspej** [comment]
> @ewerspej funny... I have the opposite opinion. 
> 
> You don't need this option for desktop, since google drive has a "make available offline" option. When you select this for a specific folder, google drive will keep a local copy on your disk.
> on the other side, a cache is necessary on mobile if you find yourself offline. No data in a foreign country, out of coverage, etc.

"Make available offline" requires you to know where the files are located. You would need to know the encrypted names of all the files and folders in question in order to make them available offline in the Cryptomator vault.

While I see some uses cases for offline mobile access, lack of data in a foreign country isn't one, because there are solutions for that. If you really need files locally on a mobile device, you can always use locally encrypted folders.

**Comment by wittymap** [comment]
> While I see some uses cases for offline mobile access, lack of data in a foreign country isn't one, because there are solutions for that (apart from WiFi, you can get an eSIM, e.g. using the Saily app). If you really need files locally on a mobile device, you can always use locally encrypted folders.
Could you say more about what you mean by locally encrypted folders? I frequently need this feature while at work. I work in an office that happens to have bad cell phone service. Cryptomator (via Dropbox) is blocked somehow on my company's guest wifi (the only network I can connect a personal device to). Currently, I have to walk outside nearly 500m down the block to find good cell coverage and get files. It's very frustrating.

**Comment by soberdor** [comment]
I agree this is not too urgent for desktop, as it can be taken care of by the cloud provider (as long as you are ready to cache the complete vault).

For mobile, it can be as simple as „no cellular coverage in this room“, or also roaming limitations - here the missing offline-option becomes a real limitation. I experienced it multiple times already and would really love to see this additional flexibility. Due to the more constrained storage situation on mobile devices, it would be great to have the option to pre-cache on folder level.


(sent from my mobile device)

Am 01.01.2025 um 21:16 schrieb Julian Ewers-Peters ***@***.***>:

﻿

@ewerspej<https://github.com/ewerspej> funny... I have the opposite opinion.

You don't need this option for desktop, since google drive has a "make available offline" option. When you select this for a specific folder, google drive will keep a local copy on your disk.
on the other side, a cache is necessary on mobile if you find yourself offline. No data in a foreign country, out of coverage, etc.

"Make available offline" requires you to know where the files are located. You would need to know the encrypted names of all the files and folders in question in order to make them available offline in the Cryptomator vault.

While I see some uses cases for offline mobile access, lack of data in a foreign country isn't one, because there are solutions for that. If you really need files locally on a mobile device, you can always use locally encrypted folders.

**Comment by ewerspej** [comment]
> Could you say more about what you mean by locally encrypted folders?

This solution won't work with Cryptomator. 

You could use file encryption tools that work locally only without cloud integration. That way, you would keep an encrypted copy on your mobile device.

On iOS you could use tools like Folder Lock and on Android something like Crypt4All apart from native on-board solutions (Samsung's flavor of Android comes with a secure folder).

**Comment by soberdor** [comment]
I don‘t understand how this should replace the desired functionality? It‘s about encrypted data, locally available and synced across multiple devices at the same time.

**Comment by fiore42** [comment]
> "Make available offline" requires you to know where the files are located. You would need to know the encrypted names of all the files and folders in question in order to make them available offline in the Cryptomator vault.

@ewerspej my apologies - I did not clarify that I assumed on desktop I have enough space to keep offline the entire vault.

currently the only option I see on mobile is creating two vaults - one on cloud and one stored locally with the file I always want with me. Very very suboptimal.

I hope this thread shows interest from users and put a conversation about this feature back on the agenda.

**Comment by ewerspej** [comment]
> @ewerspej my apologies - I did not clarify that I assumed on desktop I have enough space to keep offline the entire vault.

While that may be the case for you, this cannot serve as a generalized assumption. There are scenarios where you only want specific files or folders to be available offline as is the current functionality of existing cloud storage solutions already.

An encryption software that claims to be a replacement for Boxcryptor must also tap into this functionality otherwise it's simply not a valid replacement and shouldn't be advertised as such
